<a href="https://colab.research.google.com/github/Rafia-Shaikh-eng/Data-Science-Analytics/blob/main/1_csv_exercise_solution_ETL_Extract.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Goal of the ETL Lesson

One goal of this ETL pipelines lesson is to take the [World Bank Project data set](https://datacatalog.worldbank.org/dataset/world-bank-projects-operations) and merge this data with the [World Bank indicator data](https://data.worldbank.org/indicator/SP.POP.TOTL). Then you'll load the merged data into a database.

In the process, you'll need to transform these data sets in different ways. And finally, you'll write a single Python module that reads in these date sets, transforms them, and loads the results into the database all in one step.

# Extracting data from a csv file

The first step in an ETL pipeline is extraction. Data comes in all types of different formats, and you'll practice extracting data from csv files, json files, xml files, SQL databases, and the web.

In this first exercise, you'll practice extracting data from a CSV file and then navigating through the results. You'll see that extracting data is not always a straight-forward process.

# Part 1 projects_data.csv

There are two csv files loaded into this workspace:
* projects_data.csv
* population_data.csv

As a first step, try importing the projects data using the pandas [read_csv method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). The file path is just 'projects_data.csv'. You can see the file if you click on File->Open in the workspace.

In [1]:
## import the projects_data.csv file using the pandas library
import pandas as pd
df_projects = pd.read_csv('/content/projects_data.csv')

<ipython-input-1-8e3a9c096338>:3: DtypeWarning: Columns (44) have mixed types. Specify dtype option on import or set low_memory=False.
  df_projects = pd.read_csv('/content/projects_data.csv')


Did you get a DType warning? Read about what this warning is in the [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.errors.DtypeWarning.html).

Pandas could not automatically figure out the data type for each column (ie integer, string, etc.). That is because some columns have multiple data types. In other words, this data is somewhat messy.

Try reading in the data set again using the read_csv() method. This time, also use the option dtype=str so that pandas treats everything like a string.

In [2]:
# Read in the projects_data.csv file using the read_csv method
# and dtype = str option
df_projects = pd.read_csv('/content/projects_data.csv', dtype=str)

In [3]:
# Run the cell below to see what the data looks like
df_projects.head()

id  regionname                                        countryname  \
0  P162228       Other                                        World;World   
1  P163962      Africa  Democratic Republic of the Congo;Democratic Re...   
2  P167672  South Asia  People's Republic of Bangladesh;People's Repub...   
3  P158768  South Asia  Islamic  Republic of Afghanistan;Islamic  Repu...   
4  P161364      Africa  Federal Republic of Nigeria;Federal Republic o...   

  prodline                  lendinginstr lendinginstrtype  \
0       RE  Investment Project Financing               IN   
1       PE  Investment Project Financing               IN   
2       PE  Investment Project Financing               IN   
3       PE  Investment Project Financing               IN   
4       PE  Investment Project Financing               IN   

  envassesmentcategorycode supplementprojectflg productlinetype  \
0                        C                    N               L   
1                        B                    N               L   
2                      NaN                    Y               L   
3                        A                    N               L   
4                        B                    N               L   

  projectstatusdisplay  ... mjtheme3name mjtheme4name mjtheme5name  \
0               Active  ...          NaN          NaN          NaN   
1               Active  ...          NaN          NaN          NaN   
2               Active  ...          NaN          NaN          NaN   
3               Active  ...          NaN          NaN          NaN   
4               Active  ...          NaN          NaN          NaN   

                                            location  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  0002327546!$!Ogun State!$!7!$!3.58333!$!NG;000...   

                                            GeoLocID  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  0002327546;0002328925;0002565340;0002565343;00...   

                                          GeoLocName                 Latitude  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3                                                NaN                      NaN   
4  Ogun State;Niger State;Abia State;Edo;Kebbi St...  7;10;5.41667;6.5;11.5;8   

                Longitude            Country Unnamed: 56  
0                     NaN                NaN         NaN  
1                     NaN                NaN         NaN  
2                     NaN                NaN         NaN  
3                     NaN                NaN         NaN  
4  3.58333;6;7.5;6;4;10.5  NG;NG;NG;NG;NG;NG         NaN  

[5 rows x 57 columns]

In [4]:
# TODO: count the number of null values in the data set
# use the isnull() and sum() methods
df_projects.isnull().sum()

id                              0
regionname                      0
countryname                     0
prodline                        0
lendinginstr                  246
lendinginstrtype              246
envassesmentcategorycode     5811
supplementprojectflg           53
productlinetype                 0
projectstatusdisplay            4
status                          4
project_name                    0
boardapprovaldate            1504
board_approval_month         1504
closingdate                  3349
lendprojectcost               125
ibrdcommamt                     0
idacommamt                      0
totalamt                        0
grantamt                        0
borrower                     5919
impagency                    6097
url                             0
projectdoc                  18248
majorsector_percent         18248
sector1                         0
sector2                      8721
sector3                     11761
sector4                     13872
sector5       

Notice that the number 18248 shows up multiple times. There is also a countryname column with 0 missing values and a Country column with 14045 missing values. This data set has some issues that will need to be solved in the transform part of the pipeline.

Next, output the size of the data frame.

In [6]:
# TODO: output the shape of the data frame
df_projects.shape

(18248, 57)

The number 18248 is interesting. Many columns in the data set are filled with only NaN values.

# Part 2 population_data.csv

Next, use the pandas read_csv method to read in the population_data.csv file. The path to this file is "population_data.csv". You'll see that the results are not as expected.

In [7]:
# TODO: read in the population_data.csv file
df_population = pd.read_csv('/content/population_data.csv')

ParserError: ignored

There is something wrong with this data set. You should see an error should say "expected 3 fields in line 5, saw 63". What might have happened? Try printing out the first few lines of the data file to see what the issue might be.

In [8]:
# TODO: Print out the first 10 lines of the data set, line by line.
# HINT: You can't use the read_csv method from pandas
# HINT: to do this manually, you'll need to use pure Python
# HINT: the open(), readline(), and close() methods should be helpful
# HINT: Use a for loop
f = open('/content/population_data.csv')
for i in range(10):
    line = f.readline()
    print('line: ', i,  line)
f.close()

line:  0 "Data Source","World Development Indicators",

line:  1 

line:  2 "Last Updated Date","2018-06-28",

line:  3 

line:  4 "Country Name","Country Code","Indicator Name","Indicator Code","1960","1961","1962","1963","1964","1965","1966","1967","1968","1969","1970","1971","1972","1973","1974","1975","1976","1977","1978","1979","1980","1981","1982","1983","1984","1985","1986","1987","1988","1989","1990","1991","1992","1993","1994","1995","1996","1997","1998","1999","2000","2001","2002","2003","2004","2005","2006","2007","2008","2009","2010","2011","2012","2013","2014","2015","2016","2017",

line:  5 "Aruba","ABW","Population, total","SP.POP.TOTL","54211","55438","56225","56695","57032","57360","57715","58055","58386","58726","59063","59440","59840","60243","60528","60657","60586","60366","60103","59980","60096","60567","61345","62201","62836","63026","62644","61833","61079","61032","62149","64622","68235","72504","76700","80324","83200","85451","87277","89005","90853","92898","94

The first four lines in the file are not properly formatted and don't contain data. Next, read in the data using the read_csv method. But this time, use the skiprows option.

In [19]:
# TODO: read in population data skipping the first four rows
df_population = pd.read_csv('/content/projects_data.csv', skiprows=4)

<ipython-input-19-95d834f70632>:2: DtypeWarning: Columns (44) have mixed types. Specify dtype option on import or set low_memory=False.
  df_population = pd.read_csv('/content/projects_data.csv', skiprows=4)


In [20]:
# TODO: Run this cell to see what the data looks like
df_population.head()

P158768                    South Asia  \
0  P161364                        Africa   
1  P161483  Middle East and North Africa   
2  P161885                        Africa   
3  P162069                        Africa   
4  P163540                        Africa   

  Islamic  Republic of Afghanistan;Islamic  Republic of Afghanistan  PE  \
0  Federal Republic of Nigeria;Federal Republic o...                 PE   
1            Republic of Tunisia;Republic of Tunisia                 PE   
2  Federal Republic of Nigeria;Federal Republic o...                 PE   
3  Federal Republic of Nigeria;Federal Republic o...                 PE   
4  Federal Republic of Nigeria;Federal Republic o...                 PE   

   Investment Project Financing  IN    A  N  L  Active  ... Unnamed: 47  \
0  Investment Project Financing  IN    B  N  L  Active  ...         NaN   
1    Development Policy Lending  AD  NaN  N  L  Active  ...         NaN   
2  Investment Project Financing  IN    B  N  L  Active  ...         NaN   
3  Investment Project Financing  IN    B  N  L  Active  ...         NaN   
4  Investment Project Financing  IN    C  N  L  Active  ...         NaN   

  Unnamed: 48 Unnamed: 49                                        Unnamed: 50  \
0         NaN         NaN  0002327546!$!Ogun State!$!7!$!3.58333!$!NG;000...   
1         NaN         NaN                                                NaN   
2         NaN         NaN                                                NaN   
3         NaN         NaN                                                NaN   
4         NaN         NaN                                                NaN   

                                         Unnamed: 51  \
0  0002327546;0002328925;0002565340;0002565343;00...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                         Unnamed: 52              Unnamed: 53  \
0  Ogun State;Niger State;Abia State;Edo;Kebbi St...  7;10;5.41667;6.5;11.5;8   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3                                                NaN                      NaN   
4                                                NaN                      NaN   

              Unnamed: 54        Unnamed: 55 Unnamed: 56  
0  3.58333;6;7.5;6;4;10.5  NG;NG;NG;NG;NG;NG         NaN  
1                     NaN                NaN         NaN  
2                     NaN                NaN         NaN  
3                     NaN                NaN         NaN  
4                     NaN                NaN         NaN  

[5 rows x 57 columns]

Make sure to scroll over to see what the last column looks like. That last column, called 'Unnamed: 62', doesn't look very useful and is filled with NaN values.

In [21]:
# TODO: Count the number of null values in each column
df_population.isnull().sum()

P158768                                                                   0
South Asia                                                                0
Islamic  Republic of Afghanistan;Islamic  Republic of Afghanistan         0
PE                                                                        0
Investment Project Financing                                            246
IN                                                                      246
A                                                                      5810
N                                                                        53
L                                                                         0
Active                                                                    4
Active.1                                                                  4
Public-Private Partnerships and Public Investment Advisory Project        0
2018-06-27T00:00:00Z                                                   1504
June        

It looks like every year column has at least one NaN value.

In [22]:
# TODO: Sum the null values by column
# HINT: In the sum method, use axis=1
df_population.isnull().sum(axis=1)

0        23
1        33
2        34
3        33
4        34
         ..
18239    33
18240    33
18241    33
18242    33
18243    33
Length: 18244, dtype: int64

And it looks like almost every row has only one null value. That is probably from the 'Unnamed: 62' column that doesn't have any relevant information in it. Next, drop the 'Unnamed: 62' column from the data frame.

In [23]:
df_population.columns

Index(['P158768', 'South Asia',
       'Islamic  Republic of Afghanistan;Islamic  Republic of Afghanistan',
       'PE', 'Investment Project Financing', 'IN', 'A', 'N', 'L', 'Active',
       'Active.1',
       'Public-Private Partnerships and Public Investment Advisory Project',
       '2018-06-27T00:00:00Z', 'June', '2023-06-28T00:00:00Z', '50,000,000',
       '0', '20,000,000', '20,000,000.1', '0.1', 'IIST', 'IIST.1',
       'http://projects.worldbank.org/P158768?lang=en', 'Unnamed: 23',
       'Unnamed: 24', '!$!0', 'Unnamed: 26', 'Unnamed: 27', 'Unnamed: 28',
       'Unnamed: 29', 'Unnamed: 30', 'Unnamed: 31', 'Unnamed: 32',
       'Unnamed: 33', 'Unnamed: 34', 'Unnamed: 35', 'Unnamed: 36', '!$!0.1',
       'Unnamed: 38', 'Unnamed: 39', 'Unnamed: 40', 'Unnamed: 41',
       'Unnamed: 42', 'Unnamed: 43', 'Unnamed: 44', 'Unnamed: 45',
       'Unnamed: 46', 'Unnamed: 47', 'Unnamed: 48', 'Unnamed: 49',
       'Unnamed: 50', 'Unnamed: 51', 'Unnamed: 52', 'Unnamed: 53',
       'Unnamed: 5

In [24]:
# TODO: drop the 'Unnamed: 62' column from the data frame,
# and save the results in the df_population variable

df_population = df_population.drop(['http://projects.worldbank.org/P158768?lang=en', 'Unnamed: 23',
       'Unnamed: 24', '!$!0', 'Unnamed: 26', 'Unnamed: 27', 'Unnamed: 28',
       'Unnamed: 29', 'Unnamed: 30', 'Unnamed: 31', 'Unnamed: 32',
       'Unnamed: 33', 'Unnamed: 34', 'Unnamed: 35', 'Unnamed: 36', '!$!0.1',
       'Unnamed: 38', 'Unnamed: 39', 'Unnamed: 40', 'Unnamed: 41',
       'Unnamed: 42', 'Unnamed: 43', 'Unnamed: 44', 'Unnamed: 45',
       'Unnamed: 46', 'Unnamed: 47', 'Unnamed: 48', 'Unnamed: 49',
       'Unnamed: 50', 'Unnamed: 51', 'Unnamed: 52', 'Unnamed: 53',
       'Unnamed: 54', 'Unnamed: 55', 'Unnamed: 56'], axis=1)

In [25]:
# TODO: Now output any row that contains a null value.
# HINT: Use the isnull() and any() methods. You'll want to use axis=1
# in the any method.
df_population[df_population.isnull().any(axis=1)]

P158768                    South Asia  \
0      P161364                        Africa   
1      P161483  Middle East and North Africa   
2      P161885                        Africa   
3      P162069                        Africa   
4      P163540                        Africa   
...        ...                           ...   
18239  P006578   Latin America and Caribbean   
18240  P037451       Europe and Central Asia   
18241  P037362       Europe and Central Asia   
18242  P037452       Europe and Central Asia   
18243  P037383       Europe and Central Asia   

      Islamic  Republic of Afghanistan;Islamic  Republic of Afghanistan  PE  \
0      Federal Republic of Nigeria;Federal Republic o...                 PE   
1                Republic of Tunisia;Republic of Tunisia                 PE   
2      Federal Republic of Nigeria;Federal Republic o...                 PE   
3      Federal Republic of Nigeria;Federal Republic o...                 PE   
4      Federal Republic of Nigeria;Federal Republic o...                 PE   
...                                                  ...                 ..   
18239                Republic of Chile;Republic of Chile                 PE   
18240  Grand Duchy of Luxembourg;Grand Duchy of Luxem...                 PE   
18241              Kingdom of Denmark;Kingdom of Denmark                 PE   
18242  Kingdom of the Netherlands;Kingdom of the Neth...                 PE   
18243                    French Republic;French Republic                 PE   

       Investment Project Financing  IN    A  N  L  Active  ...  \
0      Investment Project Financing  IN    B  N  L  Active  ...   
1        Development Policy Lending  AD  NaN  N  L  Active  ...   
2      Investment Project Financing  IN    B  N  L  Active  ...   
3      Investment Project Financing  IN    B  N  L  Active  ...   
4      Investment Project Financing  IN    C  N  L  Active  ...   
...                             ...  ..  ... .. ..     ...  ...   
18239      Specific Investment Loan  IN  NaN  N  L  Closed  ...   
18240    Structural Adjustment Loan  AD  NaN  N  L  Closed  ...   
18241    Structural Adjustment Loan  AD  NaN  N  L  Closed  ...   
18242    Structural Adjustment Loan  AD  NaN  N  L  Closed  ...   
18243    Structural Adjustment Loan  AD  NaN  N  L  Closed  ...   

       2018-06-27T00:00:00Z    June  2023-06-28T00:00:00Z   50,000,000  \
0      2018-06-27T00:00:00Z    June  2023-05-31T00:00:00Z  100,000,000   
1      2018-06-27T00:00:00Z    June  2019-12-31T00:00:00Z  500,000,000   
2      2018-06-27T00:00:00Z    June  2023-10-31T00:00:00Z  765,000,000   
3      2018-06-27T00:00:00Z    June  2023-12-31T00:00:00Z  232,000,000   
4      2018-06-27T00:00:00Z    June  2022-12-31T00:00:00Z  125,000,000   
...                     ...     ...                   ...          ...   
18239  1948-03-25T00:00:00Z   March  1954-12-31T00:00:00Z   13,500,000   
18240  1947-08-28T00:00:00Z  August  1949-03-31T00:00:00Z   12,000,000   
18241  1947-08-22T00:00:00Z  August  1949-03-31T00:00:00Z   40,000,000   
18242  1947-08-07T00:00:00Z  August  1948-06-30T00:00:00Z  195,000,000   
18243  1947-05-09T00:00:00Z     May  1947-12-31T00:00:00Z  250,000,000   

                 0   20,000,000 20,000,000.1 0.1   IIST IIST.1  
0                0  100,000,000  100,000,000   0    NaN    NaN  
1      500,000,000            0  500,000,000   0    EET   IIST  
2                0  350,000,000  350,000,000   0    NaN    NaN  
3                0  225,000,000  225,000,000   0  CIATI    NaN  
4                0  125,000,000  125,000,000   0    NaN    NaN  
...            ...          ...          ...  ..    ...    ...  
18239   13,500,000            0   13,500,000   0    NaN    NaN  
18240   12,000,000            0   12,000,000   0    NaN    NaN  
18241   40,000,000            0   40,000,000   0    NaN    NaN  
18242  195,000,000            0  195,000,000   0    NaN    NaN  
18243  250,000,000            0  250,000,000   0    NaN    NaN  

[10140